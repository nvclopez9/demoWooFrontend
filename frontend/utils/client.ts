import {
  wpNonceHash,
  time,
  HOUR_IN_SECONDS,
  MINUTE_IN_SECONDS,
  DAY_IN_SECONDS,
} from '@woographql/utils/nonce';

type Creds = {
  userAgent: string;
  ip: string;
  issued: number;
}

/**
 * Creates a Client Session ID and saves it to sessionStorage.
 * 
 * @returns {string} Client Session ID  
 * @returns {string} Client Session ID expiration timestamp
 */
async function createClientSessionId() {
  const encodedCredentials = localStorage.getItem(process.env.CLIENT_CREDENTIALS_LS_KEY as string);
  let credentials: null|Creds = encodedCredentials ? JSON.parse(encodedCredentials) : null;
  if (!credentials || time() > credentials.issued + (14 * DAY_IN_SECONDS)) {
    // Create credentials object with UserAgent.
    credentials = {
      userAgent: window?.navigator?.userAgent || '',
      ip: '',
      issued: 0,
    };

    // Fetch IP.
    const response = await fetch('https://api.ipify.org/?format=json');
    const { data } = await response.json();
    credentials.ip = data?.ip || '';
  }

  // Update timestamp to ensure new nonces are generated everytime
  // the end-user starts that application.
  credentials.issued = time();
  localStorage.setItem(process.env.CLIENT_CREDENTIALS_LS_KEY as string, JSON.stringify(credentials));

  // Generate Client Session ID.
  const clientSessionId = wpNonceHash(JSON.stringify(credentials));
  const timeout = `${credentials.issued + HOUR_IN_SECONDS}`;

  // Save Client Session ID.
  sessionStorage.setItem(process.env.CLIENT_SESSION_SS_KEY as string, clientSessionId);
  sessionStorage.setItem(process.env.CLIENT_SESSION_EXP_SS_KEY as string, timeout);

  // Return Client Session ID.
  return { clientSessionId, timeout };
}

/**
 * Checks if a session token exists in localStorage.
 * 
 * @returns {string} Session Token
 */
function hasSessionToken() {
  const sessionToken = localStorage.getItem(process.env.SESSION_TOKEN_LS_KEY as string);

  return !!sessionToken;
}

let clientSetter: ReturnType<typeof setInterval>;

/**
 * Creates timed fetcher for renewing client credentials and client session id.
 * 
 * @returns {void}
 */
function setClientFetcher() {
  if (clientSetter) {
    clearInterval(clientSetter);
  }
  clientSetter = setInterval(
    async () => {
      if (!hasSessionToken()) {
        clearInterval(clientSetter);
        return;
      }
      createClientSessionId();
    },
    Number(45 * MINUTE_IN_SECONDS),
  );
}

/**
 * Returns Client Session ID and expiration timestamp.
 * 
 * @returns {string} Client Session ID
 * @returns {string} Client Session ID expiration timestamp
 */
export async function getClientSessionId() {
  let clientSessionId = sessionStorage.getItem(process.env.CLIENT_SESSION_SS_KEY as string);
  let timeout = sessionStorage.getItem(process.env.CLIENT_SESSION_EXP_SS_KEY as string);
  if (!clientSessionId || !timeout || time() > Number(timeout)) {
    ({ clientSessionId, timeout } = await createClientSessionId());
    setClientFetcher();
  }

  return { clientSessionId, timeout };
}

export function deleteClientSessionId() {
  if (clientSetter) {
    clearInterval(clientSetter);
  }
  sessionStorage.removeItem(process.env.CLIENT_SESSION_SS_KEY as string);
  sessionStorage.removeItem(process.env.CLIENT_SESSION_EXP_SS_KEY as string);
}

export function deleteClientCredentials() {
  deleteClientSessionId();
  localStorage.removeItem(process.env.CLIENT_CREDENTIALS_LS_KEY as string);
}