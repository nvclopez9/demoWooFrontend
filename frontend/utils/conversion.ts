export const emStringToPixels = (emString: string) => {
  /*
    * params:
    *  emString: a string in the form of "XXem"
    * return:
    *  number of pixels in that em string, as a number
    */
  const basePixels = 16;
  const ems = emString.split('em')[0];
  return Number(ems) * basePixels;
};