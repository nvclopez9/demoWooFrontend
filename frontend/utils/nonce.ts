import HmacMD5 from 'crypto-js/hmac-md5';
import jwtDecode from 'jwt-decode';

export const MINUTE_IN_SECONDS = 60;
export const HOUR_IN_SECONDS = 60 * MINUTE_IN_SECONDS;
export const DAY_IN_SECONDS = 24 * HOUR_IN_SECONDS;

export function time() {
  return Math.floor(new Date().getTime() / 1000);
}

export function nonceTick() {
  const nonceLife = DAY_IN_SECONDS;
  return Math.ceil(time() / (nonceLife / 2));
}

export function wpNonceHash(data: string) {
  const nonceSalt = process.env.NONCE_KEY as string + process.env.NONCE_SALT as string;
  const hash = HmacMD5(data, nonceSalt).toString();

  return hash;
}

export function createNonce(action: string, uId:string|number, token:string) {
  const i = nonceTick();

  const nonce = wpNonceHash(`${i}|${action}|${uId}|${token}`).slice(-12, -2);

  return nonce;
}

export enum ActionTypes {
  Cart = 'cart',
  Checkout = 'checkout',
  Account = 'account',
}

function getAction(action: ActionTypes, uId: string|number) {
  switch (action) {
    case ActionTypes.Cart:
      return `load-cart_${uId}`;
    case ActionTypes.Checkout:
      return `load-checkout_${uId}`;
    case ActionTypes.Account:
      return `load-account_${uId}`;
    default:
      throw new Error('Invalid nonce action provided.');
  }
}

function getNonceParam(action: ActionTypes) {
  switch (action) {
    case ActionTypes.Cart:
      return '_wc_cart';
    case ActionTypes.Checkout:
      return '_wc_checkout';
    case ActionTypes.Account:
      return '_wc_account';
    default:
      throw new Error('Invalid nonce action provided.');
  }
}

type DecodedToken = {
  data: { customer_id: string };
}
export function getUidFromToken(sessionToken: string) {
  const decodedToken = jwtDecode<DecodedToken>(sessionToken);
  if (!decodedToken?.data?.customer_id) {
    throw new Error('Failed to decode session token');
  }
  return decodedToken.data.customer_id;
}

export function generateUrl(sessionToken:string, clientSessionId:string, actionType: ActionTypes) {
  const uId = getUidFromToken(sessionToken);
  const action = getAction(actionType, uId);

  // Create nonce
  const nonce = createNonce(action, uId, clientSessionId);

  // Create URL.
  const param = getNonceParam(actionType);
  let url = `${process.env.BACKEND_URL}/wp/transfer-session?session_id=${uId}&${param}=${nonce}`;

  return url;
}