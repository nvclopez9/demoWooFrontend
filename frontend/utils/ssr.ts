export const isBrowser = () => typeof window !== 'undefined';
export const isNavigator = () => typeof window !== 'undefined' && typeof window.navigator !== 'undefined';
export function isSSR() {
  return typeof window === 'undefined';
}