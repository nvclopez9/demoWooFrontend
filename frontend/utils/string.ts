export function ucwords(str: string) {
  return str
    .toLowerCase()
    .replace(/(^([a-zA-Z\p{M}]))|([ -][a-zA-Z\p{M}])/g,
      function(s){
        return s.toUpperCase();
    });
};

export function ucfirst(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}