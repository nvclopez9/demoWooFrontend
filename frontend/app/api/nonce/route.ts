import { NextResponse } from 'next/server';

import { UpdateSessionDocument, UpdateSessionMutation, getClient } from '@woographql/graphql';
import { ActionTypes, generateUrl } from '@woographql/utils/nonce';

type RequestBody = {
  sessionToken: string;
  authToken?: string;
  clientSessionId: string;
  timeout: number;
}

type GraphQLRequestHeaders = {
  'woocommerce-session': string;
  Authorization?: string;
}

export async function POST(request: Request) {
  try {
    const {
      sessionToken,
      authToken,
      clientSessionId,
      timeout,
    } = await request.json() as RequestBody;

    if (!sessionToken) {
      return NextResponse.json({ errors: { message: 'No session started.' } }, { status: 500 });
    }

    if (!clientSessionId || !timeout) {
      return NextResponse.json({ errors: { message: 'Client Session ID and expiration must be provided.' } }, { status: 500 });
    }

    const headers: GraphQLRequestHeaders = { 'woocommerce-session': `Session ${sessionToken}` };
    if (authToken) {
      headers.Authorization = `Bearer ${authToken}`;
    }
    const client = getClient();
    client.setHeaders(headers);

    const input = {
      sessionData: [
        {
          key: 'client_session_id',
          value: clientSessionId,
        },
        {
          key: 'client_session_id_expiration',
          value: timeout,
        },
      ],
    };
    const results = await client.request<UpdateSessionMutation>(
      UpdateSessionDocument,
      { input },
    );

    if (!results.updateSession || !results.updateSession.customer) {
      const message = 'Failed to update session';
      return NextResponse.json({ errors: { message } }, { status: 500 });
    }

    const {
      cartUrl,
      checkoutUrl,
      accountUrl,
    } = results.updateSession.customer;

    // const cartUrl = generateUrl(sessionToken, clientSessionId, ActionTypes.Cart);
    // const checkoutUrl = generateUrl(sessionToken, clientSessionId, ActionTypes.Checkout);
    // const accountUrl = generateUrl(sessionToken, clientSessionId, ActionTypes.Account);

    return NextResponse.json({ cartUrl, checkoutUrl, accountUrl });
  } catch (err) {
    console.log(err);
    return NextResponse.json({ errors: { message: 'Sorry, something went wrong' } }, { status: 500 });
  }
}
