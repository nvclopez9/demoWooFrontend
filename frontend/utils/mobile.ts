'use client'
/* istanbul ignore file */
// TODO: add test coverage
import { isBrowser } from '@woographql/utils/ssr';
import { emStringToPixels } from '@woographql/utils/conversion';

export const getWindowWidth = (): number => window.innerWidth
  || document.documentElement.clientWidth
  || document.body.clientWidth;

export const getWindowHeight = (): number => window.innerHeight
  || document.documentElement.clientHeight
  || document.body.clientHeight;

const breakpointToPixels = (breakpoint: string): string|number => {
  let breakpointPixels;
  if (breakpoint.endsWith('em')) {
    breakpointPixels = emStringToPixels(breakpoint);
  } else {
    [breakpointPixels] = breakpoint.split('px');
  }

  return breakpointPixels;
};

export const isLessThanBreakpoint = (breakpoint: string): boolean => {
  /*
   * Determines whether the current window viewport is less
   * than a given breakpoint, in ems.  Uses ems because that's how
   * our breakpoints are currently defined in `theme.ts`
   *
   * params:
   *  - breakpoint: string of the form 'XXem'
   * returns
   *  - boolean: true if less than the given breakpoint, otherwise false
   */
  if (!isBrowser()) return false; // SSR sanity check

  const breakpointPixels = breakpointToPixels(breakpoint);

  const width = getWindowWidth();
  return width < Number(breakpointPixels);
};

export const isMoreThanBreakpoint = (breakpoint: string): boolean => {
  /*
   * Determines whether the current window viewport is less
   * than a given breakpoint, in ems.  Uses ems because that's how
   * our breakpoints are currently defined in `theme.ts`
   *
   * params:
   *  - breakpoint: string of the form 'XXem'
   * returns
   *  - boolean: true if less than the given breakpoint, otherwise false
   */
  if (!isBrowser()) return false; // SSR sanity check

  const breakpointPixels = breakpointToPixels(breakpoint);

  const width = getWindowWidth();
  return width >= Number(breakpointPixels);
};