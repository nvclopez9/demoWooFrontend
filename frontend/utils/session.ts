import { GraphQLError } from 'graphql/error';

import { Customer, Cart } from '@woographql/graphql';
import { isSSR } from '@woographql/utils/ssr';
import { getClientSessionId } from '@woographql/utils/client';
import { MINUTE_IN_SECONDS, time } from '@woographql/utils/nonce';


export const isDev = () => !!process.env.WEBPACK_DEV_SERVER;

// Fetch
type ResponseErrors = {
  errors?: {
    message: string;
    data?: unknown;
  }
}
async function apiCall<T>(url: string, input: globalThis.RequestInit) {
  const response = await fetch(
    url,
    {
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
      ...input,
    },
  );

  const json: T&ResponseErrors = await response.json();

  // Capture errors.
  if (json?.errors || response.status !== 200) {
    if (isDev()) {
      // eslint-disable-next-line no-console
      console.error(json?.errors || response.statusText);
    }
    throw new Error(json.errors?.message || `Failed to fetch: ${url}`);
  }

  return json;
}

// Auth management.
function saveCredentials(authToken: string, sessionToken?: string, refreshToken?: string) {
  if (isSSR()) {
    return;
  }

  sessionStorage.setItem(process.env.AUTH_TOKEN_SS_KEY as string, authToken);
  if (!!sessionToken) {
    localStorage.setItem(process.env.SESSION_TOKEN_LS_KEY as string, sessionToken);
  }
  if (refreshToken) {
    localStorage.setItem(process.env.REFRESH_TOKEN_LS_KEY as string, refreshToken);
  }
}

function saveSessionToken(sessionToken: string) {
  localStorage.setItem(process.env.SESSION_TOKEN_LS_KEY as string, sessionToken);
}

export function hasCredentials() {
  if (isSSR()) {
    return false;
  }
  const sessionToken = localStorage.getItem(process.env.SESSION_TOKEN_LS_KEY as string);
  const authToken = sessionStorage.getItem(process.env.AUTH_TOKEN_SS_KEY as string);
  const refreshToken = localStorage.getItem(process.env.REFRESH_TOKEN_LS_KEY as string);

  if (!!sessionToken && !!authToken && !!refreshToken) {
    return true;
  }

  return false;
}

function setAuthTokenExpiry() {
  const authTimeout = time() + (15 * MINUTE_IN_SECONDS);
  sessionStorage.setItem(process.env.AUTH_TOKEN_EXPIRY_SS_KEY as string, `${authTimeout}`);
}

function authTokenIsExpired() {
  const authTimeout = sessionStorage.getItem(process.env.AUTH_TOKEN_EXPIRY_SS_KEY as string);
  if (!authTimeout || Number(authTimeout) < time()) {
    return true;
  }
}

type FetchAuthTokenResponse = {
  authToken: string;
  sessionToken: string;
}

async function fetchAuthToken() {
  const refreshToken = localStorage.getItem(process.env.REFRESH_TOKEN_LS_KEY as string);
  if (!refreshToken) {
    // eslint-disable-next-line no-console
    isDev() && console.error('Unauthorized');
    return null;
  }

  const json = await apiCall<FetchAuthTokenResponse>(
    '/api/auth',
    {
      method: 'POST',
      body: JSON.stringify({ refreshToken }),
    },
  );

  const { authToken, sessionToken } = json;
  saveCredentials(authToken, sessionToken);
  setAuthTokenExpiry();

  return authToken;
}

let tokenSetter: ReturnType<typeof setInterval>;
function setAutoFetcher() {
  if (tokenSetter) {
    clearInterval(tokenSetter);
  }
  tokenSetter = setInterval(
    async () => {
      if (!hasCredentials()) {
        clearInterval(tokenSetter);
        return;
      }
      fetchAuthToken();
    },
    Number(process.env.AUTH_KEY_TIMEOUT || 30000),
  );
}

type LoginResponse = {
  authToken: string
  refreshToken: string;
  sessionToken: string;
}
export async function login(username: string, password: string): Promise<boolean|string> {
  let json: LoginResponse;
  try {
    json = await apiCall<LoginResponse>(
      '/api/login',
      {
        method: 'POST',
        body: JSON.stringify({ username, password }),
      },
    );
  } catch (error) {
    return (error as GraphQLError)?.message || error as string;
  }

  const { authToken, refreshToken, sessionToken } = json;
  saveCredentials(authToken, sessionToken, refreshToken);
  setAutoFetcher();

  return true;
}

export async function getAuthToken() {
  let authToken = sessionStorage.getItem(process.env.AUTH_TOKEN_SS_KEY as string);
  if (!authToken || authTokenIsExpired()) {
    authToken = await fetchAuthToken();
  }

  if (authToken && !tokenSetter) {
    setAutoFetcher();
  }
  return authToken;
}

type FetchSessionTokenResponse = {
  sessionToken: string;
}

async function fetchSessionToken() {
  const json = await apiCall<FetchSessionTokenResponse>(
    '/api/auth',
    { method: 'GET' },
  );

  const { sessionToken } = json;

  sessionToken && saveSessionToken(sessionToken);
  return sessionToken;
}

async function getSessionToken() {
  let sessionToken = localStorage.getItem(process.env.SESSION_TOKEN_LS_KEY as string);
  if (!sessionToken) {
    sessionToken = await fetchSessionToken();
  }
  return sessionToken;
}

export function hasRefreshToken() {
  const refreshToken = localStorage.getItem(process.env.REFRESH_TOKEN_LS_KEY as string);

  return !!refreshToken;
}

export function hasAuthToken() {
  const authToken = sessionStorage.getItem(process.env.AUTH_TOKEN_SS_KEY as string);

  return !!authToken;
}

export type FetchSessionResponse = {
  customer: Customer;
  cart: Cart;
}
export async function getSession(): Promise<FetchSessionResponse|string> {
  const authToken = await getAuthToken();
  const sessionToken = await getSessionToken();
  let json: FetchSessionResponse;
  try {
    json = await apiCall<FetchSessionResponse>(
      '/api/session',
      {
        method: 'POST',
        body: JSON.stringify({
          sessionToken,
          authToken,
        }),
      },
    );
  } catch (error) {
    return (error as GraphQLError)?.message || error as string;
  }

  const { customer } = json;
  saveSessionToken(customer.sessionToken as string);

  return json;
}

type FetchCartResponse = {
  cart: Cart;
  sessionToken: string;
}
export type CartAction = {
  mutation: 'add';
  productId: number;
  quantity: number;
  variationId?: number;
  variation?: {
    attributeName: string;
    attributeValue: string;
  }[]
  extraData?: string;
} | {
  mutation: 'update';
  items: { key: string, quantity: number }[];
} | {
  mutation: 'remove';
  keys: string[];
  all?: boolean;
}

export async function updateCart(input: CartAction): Promise<Cart|string> {
  const sessionToken = await getSessionToken();
  const authToken = await getAuthToken();
  let json: FetchCartResponse;
  try {
    json = await apiCall<FetchCartResponse>(
      '/api/cart',
      {
        method: 'POST',
        body: JSON.stringify({
          sessionToken,
          authToken,
          input,
        }),
      },
    );
  } catch (error) {
    return (error as GraphQLError)?.message || error as string;
  }

  const { cart } = json;
  saveSessionToken(json.sessionToken);

  return cart;
}

export type FetchAuthURLResponse = {
  cartUrl: string;
  checkoutUrl: string;
  accountUrl: string
}
export async function fetchAuthURLs(): Promise<FetchAuthURLResponse|string> {
  const authToken = await getAuthToken();
  const sessionToken = await getSessionToken();
  const { clientSessionId, timeout } = await getClientSessionId();
  let json: FetchAuthURLResponse;
  try {
    json = await apiCall<FetchAuthURLResponse>(
      '/api/nonce',
      {
        method: 'POST',
        body: JSON.stringify({
          sessionToken,
          authToken,
          clientSessionId,
          timeout,
        }),
      },
    );
  } catch (error) {
    return (error as GraphQLError)?.message || error as string;
  }

  return json;
}

export function deleteCredentials() {
  if (isSSR()) {
    return;
  }

  if (tokenSetter) {
    clearInterval(tokenSetter);
  }
  localStorage.removeItem(process.env.SESSION_TOKEN_LS_KEY as string);
  sessionStorage.removeItem(process.env.AUTH_TOKEN_SS_KEY as string);
  localStorage.removeItem(process.env.REFRESH_TOKEN_LS_KEY as string);
}