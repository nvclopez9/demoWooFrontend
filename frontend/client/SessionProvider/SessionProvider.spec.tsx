// SessionProvider.test.tsx
import {
  render,
  fireEvent,
  waitFor,
  act,
  screen,
} from '@woographql/testing';
import { time } from '@woographql/utils/nonce';
import { SessionProvider, useSession } from '.';

// Mock child component to test useSession hook
const MockChildComponent = () => {
  const session = useSession();
  return (
    <div>
      <p>{session.isAuthenticated ? 'Authenticated' : 'Not Authenticated'}</p>
      <p>{session.hasCredentials ? 'Has Credentials' : 'No Credentials'}</p>
      <button onClick={() => session.login('test', 'test')}>Login</button>
    </div>
  );
};

// Mock fetch calls for when unauthenticated user mounts SessionProvider.
global.fetch = jest.fn((url, options) => {
  switch (true) {
    case url === '/api/auth' && options?.method === 'GET':
      return Promise.resolve({
          status: 200,
          json: () => Promise.resolve({ sessionToken: 'sessionToken' }),
        } as Response);
    case url === '/api/session':
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({ customer: {
          id: 'guest',
          firstName: 'test',
          lastName: 'test',
          sessionToken: 'sessionTokenFromCustomer',
        } }),
      } as Response)
    case url === '/api/nonce':
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({
          cartUrl: 'cartUrl',
          checkoutUrl: 'checkoutUrl',
          accountUrl: 'accountUrl',
        }),
      } as Response)
    default:
      return Promise.resolve({
        status: 500,
        json: () => Promise.resolve({ errors: 'Invalid request made'}),
      } as Response)
  }
});

describe('SessionProvider component', () => {
  const mockFetch = global.fetch as jest.Mock<Promise<Response>>;
  beforeEach(() => {
    // Mock environment variables
    process.env.CLIENT_CREDENTIALS_LS_KEY = 'testClientCredentials';
    process.env.NONCE_KEY = 'testNonce';
    process.env.NONCE_SALT = 'testNonceSalt';
    process.env.AUTH_TOKEN_SS_KEY = 'testAuthToken';
    process.env.SESSION_TOKEN_LS_KEY = 'testSessionToken';
    process.env.REFRESH_TOKEN_LS_KEY = 'testRefreshToken';
    process.env.AUTH_KEY_TIMEOUT = '30000';

    localStorage.setItem(
      process.env.CLIENT_CREDENTIALS_LS_KEY as string,
      JSON.stringify({
        userAgent: 'testAgent',
        ip: '1.1.1.1',
        issued: time(),
      })
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

  it('renders correctly', async () => {
    await act(async () => {
      render(
        <SessionProvider>
          <MockChildComponent />
        </SessionProvider>
      );
    });

    // Check that the child component is rendered with the correct text
    expect(mockFetch).toHaveBeenCalledTimes(3);
    expect(screen.getByText('Not Authenticated')).toBeInTheDocument();
    expect(screen.getByText('No Credentials')).toBeInTheDocument();
  });

  it('should match snapshot', async () => {
    let { baseElement } = render(
      <SessionProvider>
        <MockChildComponent />
      </SessionProvider>
    )
    await waitFor(() => {
      expect(mockFetch).toHaveBeenCalledTimes(3);
      expect(baseElement).toMatchSnapshot();
    });
  });

  it('updates session state on successful login', async () => {
    await act(async () => {
      render(
        <SessionProvider>
          <MockChildComponent />
        </SessionProvider>
      );
    });

    expect(mockFetch).toHaveBeenCalledTimes(3);

    mockFetch.mockImplementationOnce(() => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({
          authToken: 'authToken',
          refreshToken: 'refreshToken',
          sessionToken: 'sessionToken',
        }),
      } as Response)
    });
    mockFetch.mockImplementationOnce(() => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({ customer: {
          id: 1,
          firstName: 'test',
          lastName: 'test',
          sessionToken: 'sessionTokenFromCustomer',
        } }),
      } as Response)
    });
    mockFetch.mockImplementationOnce(() => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({
          cartUrl: 'cartUrl',
          checkoutUrl: 'checkoutUrl',
          accountUrl: 'accountUrl',
        }),
      } as Response)
    });

    const loginButton = screen.getByText('Login');
    fireEvent.click(loginButton);

    // Wait for updates
    await waitFor(() => {
      expect(mockFetch).toHaveBeenCalledTimes(6);
      expect(screen.getByText('Authenticated')).toBeInTheDocument();
      expect(screen.getByText('Has Credentials')).toBeInTheDocument();
    });
  });

  it('fetches session for returning user with refreshToken', async () => {
    // Mock localStorage to return refreshToken
    localStorage.setItem(process.env.REFRESH_TOKEN_LS_KEY as string, 'testRefreshToken');
    mockFetch.mockImplementationOnce((...params) => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({
          authToken: 'authToken',
          refreshToken: 'refreshToken',
          sessionToken: 'sessionToken',
        }),
      } as Response)
    });
    mockFetch.mockImplementationOnce((...params) => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({ customer: {
          id: 1,
          firstName: 'test',
          lastName: 'test',
          sessionToken: 'sessionTokenFromCustomer',
        } }),
      } as Response)
    });
    mockFetch.mockImplementationOnce((...params) => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve({
          cartUrl: 'cartUrl',
          checkoutUrl: 'checkoutUrl',
          accountUrl: 'accountUrl',
        }),
      } as Response)
    });

    render(
      <SessionProvider>
        <MockChildComponent />
      </SessionProvider>
    );

    // Wait for updates
    await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(3));
    expect(screen.getByText('Authenticated')).toBeInTheDocument();
    expect(screen.getByText('Has Credentials')).toBeInTheDocument();
  });
});
