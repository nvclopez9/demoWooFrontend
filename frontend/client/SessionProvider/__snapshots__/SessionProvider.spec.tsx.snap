// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionProvider component should match snapshot 1`] = `
<body>
  <div>
    <div>
      <p>
        Not Authenticated
      </p>
      <p>
        No Credentials
      </p>
      <button>
        Login
      </button>
    </div>
  </div>
</body>
`;
